lib/traiderdaive/traiderdaive/gym_environments/simple_hyperdrive_env.py

```diff
@@ -0,0 +1,443 @@
+        # Define the random bots
+        self.random_bots = [
+            self.interactive_hyperdrive.init_agent(
+                base=gym_config.random_bot_budget,
+                policy=PolicyZoo.random,
+                # TODO set the seed per random bot here for reproducability
+                policy_config=PolicyZoo.random.Config(),
+                name="random_bot_" + str(i),
+            )
+            for i in range(gym_config.num_random_bots)
+        ]
+
+        # Save a snapshot of initial conditions for resets
+        self.chain.save_snapshot()
+
+        assert gym_config.render_mode is None or gym_config.render_mode in self.metadata["render_modes"]
+        self.render_mode = gym_config.render_mode
+
+        self.gym_config = gym_config
+
+        # This action space encompasses buy and sell of bonds based on the current taken position
+        # Following https://github.com/AminHP/gym-anytrading/tree/master
+        # This binary action space with a position state variable results in one of the 4 following trades:
+        # - If the current position is short and the agent's action is to buy, the environment will
+        #     close the short position and open a long position. The current position is then set to long.
+        # - If the current position is long and the agent's action is to buy, this is noop (i.e., hold long position)
+        # - If the current position is long and the agent's action is to sell, the environment will
+        #   close the long position and open a short position. The current position is then set to short.
+        # - If the current position is short and the agent's action is to sell, this is noop (i.e., hold short position)
+        self.action_space = spaces.Discrete(len(Actions))
+
+        # The space of observations from the environment
+        # This is a timeseries of window_size samples of the spot price and the share price
+        # TODO add more data for rl bot input
+        self.observation_space = spaces.Box(low=0, high=1e10, shape=(self.gym_config.window_size, 2), dtype=np.float64)
+
+        # For a more complex environment:
+
+        # The space of allowed actions to take
+        # Following https://github.com/AminHP/gym-mtsim
+        # These actions are encoded into a 1d vector of continuous values
+        # This is due to not all algorithms supporting dict or multidimention box actions
+
+        # Here, these actions are for 3 types of trades: longs, shorts, and LP,
+        # each encoded as an array of length max_positions + 2
+        # For a given type of trade, the elements are interpreted as
+        # [
+        #    probability of closing order 1,
+        #    probability of closing order 2,
+        #    ...
+        #    probability of closing order max_positions,
+        #    probability of holding or creating a new order,
+        #    volume of the new order
```

Comment:
I'm confused about how this representation works. Wouldn't every entry need to be a probability? How are you able to have the last entry be an unbounded scalar?